packages/ckeditor5-html-support/src/integrations/dualcontent.js

```diff
@@ -113,8 +113,23 @@ export default class DualContentModelElementSupport extends Plugin {
 	_hasBlockContent( viewElement ) {
 		const blockElements = this.editor.editing.view.domConverter.blockElements;
 
-		return Array.from( viewElement.getChildren() )
-			.some( node => blockElements.includes( node.name ) );
+		return hasBlockContent( viewElement );
+
+		function isBlock( node ) {
+			if ( blockElements.includes( node.name ) ) {
+				return true;
+			}
+
+			if ( node.is( 'element' ) ) {
+				return hasBlockContent( node );
+			}
+
+			return false;
+		}
+
+		function hasBlockContent( viewElement ) {
+			return Array.from( viewElement.getChildren() ).some( isBlock );
+		}
```

Comment:
I would suggest using a `TreeWalker`:

```javascript
	_hasBlockContent( viewElement ) {
		const view = this.editor.editing.view;
		const blockElements = view.domConverter.blockElements;

		for ( const viewItem of view.createRangeIn( viewElement ).getItems() ) {
			if ( viewItem.is( 'element' ) && blockElements.includes( viewItem.name ) ) {
				return true;
			}
		}

		return false;
	}
```

It is more consistent with how we solve such issues and IMHO is easier to understand.